h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -240,28 +256,69 @@ protected MP getModelParams(MP params, Object[] hyperParams) {
-      long work = 1;
+      long work = 0;
+      long free_param_combos = 1;
+      
       for (Map.Entry<String, Object[]> p : _hyperParams.entrySet()) {
-        if (p.getValue() != null) {
-          work *= p.getValue().length;
+        if (p.getValue() != null && !p.getKey().equals("constraints")) {
+          free_param_combos *= p.getValue().length;
+        }
+      }
+      
+      if(_hyperParams.get("constraints") == null) {
+        return free_param_combos;
+      }
+      
+      for (int i = 0; i < _hyperParams.get("constraints").length; i++) {
+        long constrained_param_combos = 1;
+        Map<String, Object[]> constraints = (Map<String, Object[]>)_hyperParams.get("constraints")[i];
+        for (Map.Entry<String, Object[]> p : constraints.entrySet()) {
+          Object o = p.getValue();
+          if ((o instanceof List) && (o != null)) {
+            constrained_param_combos *= ((ArrayList) o).toArray().length;
+          }
         }
+        work += constrained_param_combos * free_param_combos;
       }
+      
       return work;
     }
 
+    protected Map<String, Object[]> mergeHashMaps(Map<String, Object[]> m1, Map<String, Object[]> m2) {
+      if(m2 == null) { return m1; }
+      Map<String, Object[]> m = new HashMap<>();
+      
+      for(Map.Entry<String, Object[]> entry : m1.entrySet()) {
+        if(entry.getKey().equals("constraints")) { continue; }
+        m.put(entry.getKey(), entry.getValue());
+      }
+      
+      for(Map.Entry<String, Object[]> entry : m2.entrySet()) {
+        Object val = entry.getValue();
+        if (val instanceof List)
+          m.put(entry.getKey(), ((ArrayList) val).toArray());
+        else {
+          Object[] singleValueArr = {val};  // hyperParams in constraint is not a list here
+          m.put(entry.getKey(), singleValueArr);
+        }
+      }
+      return m;
+    }
+    
     /** Given a list of indices for the hyperparameter values return an Object[] of the actual values. */
-    protected Object[] hypers(int[] hidx) {
-      Object[] hypers = new Object[_hyperParamNames.length];
+    protected Object[] hypers(Map<String, Object[]> hyperParams, String[] hyperParamNames, int[] hidx) {
+      Object[] hypers = new Object[hyperParamNames.length];
       for (int i = 0; i < hidx.length; i++) {
-        hypers[i] = _hyperParams.get(_hyperParamNames[i])[hidx[i]];
+        hypers[i] = hyperParams.get(hyperParamNames[i])[hidx[i]];
       }
       return hypers;
     }
 
-    protected int integerHash(int[] ar) {
-      Integer[] hashMe = new Integer[ar.length];
+    protected int integerHash(Map<String, Object[]> hyperParams, String[] hyperParamNames, int[] ar, int constraintNum) {
+      Integer[] hashMe = new Integer[ar.length + 1];
       for (int i = 0; i < ar.length; i++)
-        hashMe[i] = ar[i] * _hyperParams.get(_hyperParamNames[i]).length;
+        hashMe[i] = ar[i] * hyperParams.get(hyperParamNames[i]).length;
```

Comment:
Karthik:

In order to test this hash function implementation, can you add a test with constraints containing 5 arrays but with each array containing exactly the same model parameters?  Something like this:

{'scale': [[1, 1], [2, 2]], 'num_knots': [[5, 5], [6, 6]], 'gam_columns': [["C11", "C12"], ["C12", "C13"]]}